test/threadsafe_function/threadsafe_function_sum.cc

```diff
@@ -0,0 +1,155 @@
+  }
+  finalizeData->deferred.Resolve(Boolean::New(env,true));
+  delete finalizeData;
+}
+
+/**
+ * See threadsafe_function_sum.js for descriptions of the tests in this file
+ */
+
+void entryWithTSFN(ThreadSafeFunction tsfn, int threadId) {
+  std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 100 + 1));
+  tsfn.BlockingCall( [=](Napi::Env env, Function callback) {
+    callback.Call( { Number::New(env, static_cast<double>(threadId))});
+  });
+  tsfn.Release();
+}
+
+static Value TestWithTSFN(const CallbackInfo& info) {
+  int threadCount = info[0].As<Number>().Int32Value();
+  Function cb = info[1].As<Function>();
+
+  // We pass the test data to the Finalizer for cleanup. The finalizer is
+  // responsible for deleting this data as well.
+  TestData *testData = new TestData({
+    Promise::Deferred::New(info.Env())
+  });
+
+  ThreadSafeFunction tsfn = ThreadSafeFunction::New(
+      info.Env(), cb, "Test", 0, threadCount,
+      std::function<decltype(FinalizerCallback)>(FinalizerCallback), testData);
+
+  for (int i = 0; i < threadCount; ++i) {
+    testData->threads.push_back( thread(entryWithTSFN, tsfn, i) );
+  }
+
+  return testData->deferred.Promise();
+}
+
+
+void entryDelayedTSFN(std::future<ThreadSafeFunction> tsfnFuture, int threadId) {
+  std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 100 + 1));
+  ThreadSafeFunction tsfn = tsfnFuture.get();
+  tsfn.BlockingCall( [=](Napi::Env env, Function callback) {
+    callback.Call( { Number::New(env, static_cast<double>(threadId))});
+  });
+  tsfn.Release();
+}
+
+static Value TestDelayedTSFN(const CallbackInfo& info) {
+  int threadCount = info[0].As<Number>().Int32Value();
+  Function cb = info[1].As<Function>();
+
+  TestData *testData = new TestData({
+    Promise::Deferred::New(info.Env())
+  });
+
+  vector< std::promise<ThreadSafeFunction> > tsfnPromises;
+
+  for (int i = 0; i < threadCount; ++i) {
+    tsfnPromises.emplace_back();
+    testData->threads.push_back( thread(entryDelayedTSFN, tsfnPromises[i].get_future(), i) );
+  }
+
+  testData->tsfn = ThreadSafeFunction::New(
+      info.Env(), cb, "Test", 0, threadCount,
+      std::function<decltype(FinalizerCallback)>(FinalizerCallback), testData);
+
+  for (int i = 0; i < threadCount; ++i) {
+    tsfnPromises[i].set_value(testData->tsfn);
+  }
+
+  return testData->deferred.Promise();
+}
+
+void entryAcquire(ThreadSafeFunction tsfn, int threadId) {
+  tsfn.Acquire();
+  std::this_thread::sleep_for(std::chrono::milliseconds(std::rand() % 100 + 1));
+  tsfn.BlockingCall( [=](Napi::Env env, Function callback) {
+    callback.Call( { Number::New(env, static_cast<double>(threadId))});
+  });
+  tsfn.Release();
+}
+
+static Value CreateThread(const CallbackInfo& info) {
+  TestData* testData = static_cast<TestData*>(info.Data());
+  ThreadSafeFunction tsfn = testData->tsfn;
+  int threadId = testData->threads.size();
+  testData->threads.push_back( thread(entryAcquire, tsfn, threadId) );
```

Comment:
I wonder if it would be useful to highlight with a comment that this is where the copy constructor is being tested.  Since the main change is to add the copy constructor it would be good to make it obvious in the test that it is being used.

Reply:
Added comment